src/librustc_typeck/mem_categorization.rs

```diff
@@ -609,6 +640,62 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
         self.cat_pattern_(place, pat, &mut op)
     }
 
+    fn variant_index_for_adt(
+        &self,
+        qpath: &hir::QPath<'_>,
+        hir_id: hir::HirId,
+        span: Span,
+    ) -> Option<VariantIdx> {
+        let res = self.tables.qpath_res(qpath, hir_id);
+        let ty = self.tables.node_type(hir_id);
+        let adt_def = match ty.kind {
+            ty::Adt(adt_def, _) => adt_def,
+            _ => {
+                self.tcx()
+                    .sess
+                    .delay_span_bug(span, "struct/tuple struct pattern not applied to an ADT");
+
+                return None;
+            }
+        };
+
+        match res {
+            Res::Def(DefKind::Variant, variant_id) => {
+                Some(adt_def.variant_index_with_id(variant_id))
+            }
+            Res::Def(DefKind::Ctor(CtorOf::Variant, ..), variant_ctor_id) => {
+                Some(adt_def.variant_index_with_ctor_id(variant_ctor_id))
+            }
+            Res::Def(DefKind::Ctor(CtorOf::Struct, ..), _)
+            | Res::Def(DefKind::Struct | DefKind::Union | DefKind::TyAlias | DefKind::AssocTy, _)
+            | Res::SelfCtor(..)
+            | Res::SelfTy(..) => {
+                // Structs and Unions have only have one variant.
+                Some(VariantIdx::new(0))
+            }
+            _ => bug!("expected ADT path, found={:?}", res),
+        }
+    }
+
+    fn total_fields_in_adt_variant(
+        &self,
+        hir_id: hir::HirId,
+        variant_index: VariantIdx,
+        span: Span,
+    ) -> Option<usize> {
+        let ty = self.tables.node_type(hir_id);
+        match ty.kind {
+            ty::Adt(adt_def, _) => Some(adt_def.variants[variant_index].fields.len()),
+            _ => {
+                self.tcx()
+                    .sess
+                    .delay_span_bug(span, "struct/tuple struct pattern not applied to an ADT");
```

Comment:
do you mean struct or tuple struct?

Reply:
Yup. I can fix this -- copied it from another place lol